src/main/java/com/google/cloud/spanner/pgadapter/statements/IntermediateStatement.java

```diff
@@ -240,85 +299,191 @@ public Exception getException() {
-        }
-        for (String stmt : statements) {
-          connection.execute(Statement.of(stmt));
-        }
-        long[] updateCounts = connection.runBatch();
-        updateBatchResultCount(updateCounts);
+    executed = true;
+    if (connection.isInTransaction()) {
+      executionStatus = ExecutionStatus.EXPLICIT_TRANSACTION;
+    } else {
+      executionStatus = ExecutionStatus.AUTOCOMMIT;
+    }
+    int index = 0;
+    while (index < getStatementCount()) {
+      if (connectionHandler.getStatus() == ConnectionStatus.TRANSACTION_ABORTED
+          && !"ROLLBACK".equals(getCommand(index))
+          && !"COMMIT".equals(getCommand(index))) {
+        handleExecutionException(
+            index,
+            SpannerExceptionFactory.newSpannerException(
+                ErrorCode.INVALID_ARGUMENT,
+                "ERROR: current transaction is aborted, commands ignored until end of transaction block"));
+        index++;
+        continue;
+      }
+      if (getStatementType(index) == StatementType.DDL
+          && executionStatus == ExecutionStatus.EXPLICIT_TRANSACTION) {
+        handleExecutionExceptionAndTransactionStatus(
+            index,
+            SpannerExceptionFactory.newSpannerException(
+                ErrorCode.INVALID_ARGUMENT,
+                "ERROR: DDL statements are only allowed outside explicit transactions."));
+        index++;
+        continue;
+      }
+      boolean canUseBatch = false;
+      if (index < (getStatementCount() - 1)) {
+        StatementType statementType = getStatementType(index);
+        StatementType nextStatementType = getStatementType(index + 1);
+        canUseBatch = canBeBatchedTogether(statementType, nextStatementType);
+      }
+      if (canUseBatch) {
+        index += executeStatementsInBatch(index);
+      } else {
+        executeSingleStatement(index);
+        index++;
+      }
+    }
+  }
+
+  private void executeSingleStatement(int index) {
+    // Before executing the statement, handle specific statements that change the transaction status
+    String command = getCommand(index);
+    String statement = getStatement(index);
+    if ("BEGIN".equals(command)) {
+      if (executionStatus == ExecutionStatus.EXPLICIT_TRANSACTION) {
+        // Executing a BEGIN statement when an explicit transaction is already active is a no-op
+        return;
       } else {
-        StatementResult result =
-            connection.execute(Statement.of(this.parsedStatement.getSqlWithoutComments()));
-        updateResultCount(result);
+        executionStatus = ExecutionStatus.EXPLICIT_TRANSACTION;
       }
+    }
+    if ("COMMIT".equals(command) || "ROLLBACK".equals(command)) {
+      if (executionStatus == ExecutionStatus.AUTOCOMMIT) {
+        // Executing a COMMIT/ROLLBACK statement in autocommit mode is a no-op
+        return;
+      }
+      executionStatus = ExecutionStatus.AUTOCOMMIT;
+      if (connectionHandler.getStatus() == ConnectionStatus.TRANSACTION_ABORTED) {
```

Comment:
I'm not exactly sure how/why, but I'm not able to get out of an aborted state after a transaction has failed. Executing for example a batch like this (assuming invalid_table does not exist):

`BEGIN; INSERT INTO invalid_table VALUES (1); COMMIT;`

This will bring the connection into an aborted state, and it is impossible to get out of that state. The result of this batch should be that the transaction is rolled back by the commit, and that the connection is idle. Executing a separate rollback after the batch also does not get the connection out of the aborted state. Could you add a couple of tests for this so we know that getting out of the aborted state works.

Reply:
Ok, so the reason is that the psql interactive mode splits the statements and sends them one by one rather than as a batch. The current logic assumes that the aborted transaction cannot happen when the execution status is autocommit, which is indeed possible. I reordered those lines and it works out.